simple-request-response/README.md

```diff
@@ -0,0 +1,84 @@
+# A simple request/response abstraction
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-07-14  |
+
+Authors: [@MarcoPolo]
+
+todo: others welcome to join
+Interest Group: [@mxinden] [@marten-seemann]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@MarcoPolo]: https://github.com/MarcoPolo
+[@mxinden]: https://github.com/mxinden
+
+## Introduction
+
+Many application protocols map well to request/response semantics where a peer
+makes a request to another peer, and receives a response back. For example,
+requesting a file from a peer and getting the file back. In contrast to
+request/response, stream semantics allow two peers to continously send data
+back and forth with no specific ordering. Both styles have their uses. This
+document focuses on _one_ way of implementing request/response semantics and
+how to map it to stream based transports and an HTTP transport.
+
+## Goals of this document
+
+A goal of this document is to define _a_ simple way to write request/response
+style application protocols that can make use of all libp2p transports
+seamlessly (all stream transports and the HTTP transport). This is only one way
+of writing request/response style protocols, and users may choose to not use it.
+
+Another goal is to be backwards compatible with existing application protocols
+that follow these semantics to make upgrading them easier so that they may take
+advantage of the new HTTP transport without sacrificing backwards compatibility.
+
+The final goal is to define something simple such that it's easy to follow,
+implement, and use.
+
+## The abstraction
+
+This abstraction takes in a blob from the requesting peer (aka client) that
+represents the request and delivers the blob to the responding peer (aka
+server). The server responds with a blob that respresents the response. This
+abstraction is agnostic to what the blobs are or how they are encoded. That's up
+to the application protocol to decide.
```

Comment:
I would put this before the "goal" section such that we can extend the goal section to say that what we want is to define an interface that application protocols can rely on.

Plus, I'd perhaps reword this to:

```
## The interface

At its core, request-response means:
- taking _one_ blob from the requesting peer (aka the client)
- delivering this blob to the responding peer (aka the server)
- taking _one_ blob from the responding peer
- delivering this blob to the requesting peer

The defining characteristics are:

- Each party can only send one message
- The interaction is linear in time (the response can only be sent after the request has been received)

This is in contrast to for example streaming protocols where the responding peer may send any number of messages or event/notification-based protocols where a message may be received without a prior request.

What the blobs are and how they are encoded is subject to the application protocol and not defined in this interface.
```

I think the contrast to other communication patterns is important here:

- one message per peer vs many messages
- interaction is synchronous in nature, i.e. a client can "block" its control-flow on receiving a response

These are essentially the things we are defining that allow the creation of an interface (in code) that we can use as an abstraction for other protocols to build upon.